### PR TITLE
Support transforming range requests when origin returns full resource.

### LINKED ISSUE #8161: Revalidation on range request leads to full object being served
When a range request comes into ATS and the underlying object must be revalidated or is expired, it appears that ATS serves the entire object to the client and ignores the requested range in the `Range` header. While range requests are not required to be supported, in this case they are supported in ATS, and the difference in behavior can lead to unexpected results when the client sees the 200 and full object instead of the 206. Ideally ATS would revalidate, fill the new object as necessary, and return the requested range to the client.

### PR DESCRIPTION
**Pull request by cmcfarlen**
Resolves #8161:

This PR adds support for handling range requests after a response is received and the response source is the origin server.

 - Allow caching responses when a range is requested (requires proxy.config.http.cache.range.write is set, but slightly changes behavior)
 - Handle range setup even if the response source is the origin server.
 - Call do_range_setup_if_necessary on forward server response.
 - Enable API flag to cache original response when a partial response is tranformed (api_info.cache_untransformed = true)

This doesn't currently support transforming chunked responses from the origin.

### CODE REVIEW

**Comment by masaori335** [comment]
@ezelkow1 Would you add Chris (@cmcfarlen) to the ci list?

**Comment by vmamidi** [comment]
[approve ci]

**Comment by masaori335** [comment]
[approve ci centos]

File: tests/gold_tests/cache/cache-range-response.test.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ts = Test.MakeATSProcess("ts")
+replay_file = "replay/cache-range-response.replay.yaml"
+server = Test.MakeVerifierServerProcess("server0", replay_file)
+ts.Disk.records_config.update({
+    'proxy.config.diags.debug.enabled': 1,
+    'proxy.config.diags.debug.tags': 'http.*|cache.*',
+    'proxy.config.http.cache.range.write': 1,
+})
+ts.Disk.remap_config.AddLine(
+    'map / http://127.0.0.1:{0}'.format(server.Variables.http_port)
```
Comment (bneradt) [inline tests/gold_tests/cache/cache-range-response.test.py:33]: Small thing: we're generally using Python f stings now. Thus:

```
    f'map / http://127.0.0.1:{server.Variables.http_port}'
```

**Review by bneradt — CHANGES_REQUESTED** [review]

**Comment by bryancall** [comment]
@traeak Can you please review this PR?

**Comment by cmcfarlen** [comment]
I added a check for a Content-Encoding and skip out on setting up a server transform to a 206.

**Review by traeak — COMMENTED** [review]
Would it be possible to add this PR feature in as an option '2' for configuration's sake?

**Comment by ezelkow1** [comment]
[approve ci rocky]

**Comment by cmcfarlen** [comment]
I will check the case when a large asset is requested with just the first part specified in Range and see what happens.  Will report back.

**Comment by cmcfarlen** [comment]
I setup a scenario to try to answer the question about short ranges and caching the full asset.  TL;DR: It caches the full response

### Scenario

- https as the origin to serve a ~50MB file configured to ignore ranges to always return a 200 and the full request
- ATS mapping to origin with this new feature turned on to transform 200 to 206 and cache the full response (staring with an empty cache)
- curl client requesting to close the connection and requesting a range.

Here is the simple request script:

```
#!/bin/sh
curl -v -o first.bin  --header "Connection: close" --header 'Range: bytes=0-10' http://ats:8080/docker.iso &
sleep 0.1  # added or both requests would be cache misses
curl -v -o second.bin  --header "Connection: close" --header 'Range: bytes=10000000-' http://ats:8080/docker.iso &
```

### Expected behavior

The first request will cause ats start the transfer from origin, transforming the response and sending 11 bytes to the client and caching the full response.
The second request will do a RWW and serve the request from the cache as its written

### Results

(I removed the output related to the transfer progress, leaving just the verbose output)

```
* Connected to ats (127.0.0.1) port 8080 (#0)
> GET /docker.iso HTTP/1.1
> Host: ats:8080
> User-Agent: curl/7.77.0
> Accept: */*
> Connection: close
> Range: bytes=0-10
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 206 Partial Content
< Date: Mon, 21 Feb 2022 20:31:36 GMT
< Server: ATS/10.0.0
< Last-Modified: Mon, 21 Feb 2022 17:59:12 GMT
< ETag: "3000000-5d88afd724296"
< Accept-Ranges: none
< Cache-Control: public, max-age=60, must-revalidate
< Content-Type: application/x-iso9660-image
< Content-Range: bytes 0-10/50331648
< Content-Length: 11
< Age: 0
< Connection: close
< Via: http/1.1 traffic_server (ApacheTrafficServer/10.0.0 [uScMsSfWpSeN:t cCMp sS])
<
* Closing connection 0

* Connected to ats (127.0.0.1) port 8080 (#0)
> GET /docker.iso HTTP/1.1
> Host: ats:8080
> User-Agent: curl/7.77.0
> Accept: */*
> Connection: close
> Range: bytes=10000000-
>
$ * Mark bundle as not supporting multiuse
< HTTP/1.1 206 Partial Content
< Date: Mon, 21 Feb 2022 20:31:36 GMT
< Server: ATS/10.0.0
< Last-Modified: Mon, 21 Feb 2022 17:59:12 GMT
< ETag: "3000000-5d88afd724296"
< Accept-Ranges: none
< Cache-Control: public, max-age=60, must-revalidate
< Content-Type: application/x-iso9660-image
< Content-Range: bytes 10000000-50331647/50331648
< Content-Length: 40331648
< Age: 0
< Connection: close
< Via: http/1.1 traffic_server (ApacheTrafficServer/10.0.0 [uScWs f p eN:t cCHp s ])
<
* Closing connection 0
```

A subsequent request for the full content will serve the full resource from cache:

```
< Content-Length: 50331648
< Cache-Control: public, max-age=60, must-revalidate
< Content-Type: application/x-iso9660-image
< Age: 9
< Connection: close
< Via: http/1.1 traffic_server (ApacheTrafficServer/10.0.0 [uScHs f p eN:t cCHp s ])
```

**Comment by bneradt** [comment]
[approve ci fedora]

**Comment by traeak** [comment]
Broke this while testing with an IMS header.

**Comment by traeak** [comment]
[approve ci fedora]

**Review by traeak — APPROVED** [review]
Testing went okay, code looks okay.

**Pull request merged**